src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+        UInt       distalSegmentThreshold,
+        UInt       distalSegmentMatch,
+        UInt       distalAddSynapses,
+        Permanence distalIncrement,
+        Permanence distalDecrement,
+        Permanence distalMispredictDecrement,
+        Permanence distalSynapseThreshold,
+
+        Real stability_rate,
+        Real fatigue_rate,
+
+        Real period,
+        Int  seed,
+        bool verbose) {
+
+    proximalInputDimensions_ = proximalInputDimensions;
+    distalInputDimensions_   = distalInputDimensions;
+    inhibitionDimensions_    = inhibitionDimensions;
+    cellsPerInhibitionArea_  = cellsPerInhibitionArea;
+
+    proximalSegments_        = proximalSegments;
+    this->sparsity                    = sparsity;
+    this->proximalSegmentThreshold    = proximalSegmentThreshold;
+    this->proximalIncrement           = proximalIncrement;
+    this->proximalDecrement           = proximalDecrement;
+    this->proximalSynapseThreshold    = proximalSynapseThreshold;
+    this->stability_rate              = stability_rate;
+    this->fatigue_rate                = fatigue_rate;
+
+    SDR proximalInputs(  proximalInputDimensions );
+    SDR inhibitionAreas( inhibitionDimensions );
+    cellDimensions_ = inhibitionAreas.dimensions;
+    cellDimensions_.push_back( cellsPerInhibitionArea );
+    SDR cells( cellDimensions_ );
+    rng_ = Random(seed);
+
+    // Setup the proximal segments & synapses.
+    proximalConnections.initialize(cells.size, proximalSynapseThreshold);
+    SDR_Sparsity PP_Sp(            proximalInputs.dimensions, 10 * proximalInputs.size);
+    SDR_ActivationFrequency PP_AF( proximalInputs.dimensions, 10 * proximalInputs.size);
+    UInt cell = 0u;
+    for(auto inhib = 0u; inhib < inhibitionAreas.size; ++inhib) {
+      inhibitionAreas.setSparse(SDR_sparse_t{ inhib });
+      for(auto c = 0u; c < cellsPerInhibitionArea; ++c, ++cell) {
+        for(auto s = 0u; s < proximalSegments; ++s) {
+          auto segment = proximalConnections.createSegment( cell );
+
+          // Make synapses.
+          potentialPool( inhibitionAreas, proximalInputs );
+          for(const auto presyn : proximalInputs.getSparse() ) {
+            auto permanence = initProximalPermanence();
+            proximalConnections.createSynapse( segment, presyn, permanence);
+          }
+          proximalConnections.raisePermanencesToThreshold( segment,
+                          proximalSynapseThreshold, proximalSegmentThreshold );
+          PP_Sp.addData( proximalInputs );
+          PP_AF.addData( proximalInputs );
+        }
+      }
+    }
+    tieBreaker_.resize( proximalConnections.numSegments() );
+    for(auto i = 0u; i < tieBreaker_.size(); ++i) {
+      tieBreaker_[i] = 0.01f * rng_.getReal64();
+    }
+    proximalMaxSegment_.resize( cells.size );
+    AF = new SDR_ActivationFrequency( {cells.size, proximalSegments}, period );
+    AF->initializeToValue( sparsity / proximalSegments );
```

Comment:
ditto as PP_AF

Reply:
This AF object is used by boosting, can't delete